File: src/main/java/ru/citeck/ecos/notifications/service/processors/FirebaseNotificationProcessor.java
package ru.citeck.ecos.notifications.service.processors;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.firebase.messaging.*;
import com.rabbitmq.client.Delivery;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang.StringUtils;
import org.springframework.stereotype.Component;
import ru.citeck.ecos.events.data.dto.EventDto;
import ru.citeck.ecos.events.data.dto.pasrse.EventDtoFactory;
import ru.citeck.ecos.events.data.dto.task.TaskEventDto;
import ru.citeck.ecos.events.data.dto.task.TaskEventType;
import ru.citeck.ecos.notifications.config.ApplicationProperties;
import ru.citeck.ecos.notifications.domain.subscribe.repo.ActionEntity;
import ru.citeck.ecos.notifications.domain.subscribe.service.ActionService;
import ru.citeck.ecos.notifications.freemarker.FreemarkerTemplateEngineService;
import ru.citeck.ecos.notifications.service.TemplateService;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * @author Roman Makarskiy
 */
@Slf4j
@Component
public class FirebaseNotificationProcessor extends ActionProcessor {

    private static final String ID = "FIREBASE_NOTIFICATION";
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    private static final String ERROR_CODE_TOKEN_NOT_REGISTERED = "registration-token-not-registered";

    private static final String PARAM_FIREBASE_CLIENT_REG_TOKEN = "fireBaseClientRegToken";
    private static final String PARAM_DEVICE_TYPE = "deviceType";
    private static final String TEMPLATE_ID = "templateId";
    private static final String TITLE_TEMPLATE = "titleTemplate";
    private static final String BODY_TEMPLATE = "bodyTemplate";
    private static final String DEVICE_ANDROID = "android";
    private static final String DEVICE_IOS = "ios";
    private static final String PARAM_TASK_ID = "taskId";
    private static final String PARAM_DOCUMENT = "documentRef";
    private static final String TITLE_TEMPLATE_KEY = "titleTemplate";
    private static final String BODY_TEMPLATE_KEY = "bodyTemplate";

    private final FreemarkerTemplateEngineService templateEngineService;
    private final TemplateService templateService;
    private final ApplicationProperties appProps;
    private final ActionService actionService;

    {
        setId(ID);
    }

    public FirebaseNotificationProcessor(FreemarkerTemplateEngineService templateEngineService,
                                         TemplateService templateService, ApplicationProperties appProps,
                                         ActionService actionService) {
        this.templateEngineService = templateEngineService;
        this.templateService = templateService;
        this.appProps = appProps;
        this.actionService = actionService;
    }

    @Override
    protected void processImpl(Delivery message, EventDto dto, ActionEntity action, Map<String, Object> model) {
        log.debug("Process DTO: \n" + dto);
        JsonNode config;
        try {
            config = OBJECT_MAPPER.readValue(action.getConfigJSON(), JsonNode.class);
        } catch (IOException e) {
            throw new RuntimeException("Failed read json from string", e);
        }

        if (!config.hasNonNull(PARAM_FIREBASE_CLIENT_REG_TOKEN)) {
            throw new IllegalStateException("Action config does not contains firebase registration token");
        }

        if (!config.hasNonNull(PARAM_DEVICE_TYPE)) {
            throw new IllegalStateException("Action config does not contains device type");
        }

        String eventType = dto.resolveType();
        Template template = getTemplate(eventType, config);

        String title = templateEngineService.process(TITLE_TEMPLATE_KEY, template.getTitle(), model);
        String body = templateEngineService.process(BODY_TEMPLATE_KEY, template.getBody(), model);

        String registrationToken = config.get(PARAM_FIREBASE_CLIENT_REG_TOKEN).asText();
        String deviceType = config.get(PARAM_DEVICE_TYPE).asText();

        Message fireBaseMessage;
        Map<String, String> customData = prepareCustomData(dto);

        switch (deviceType) {
            case DEVICE_ANDROID: {
                fireBaseMessage = buildAndroidMessage(title, body, registrationToken, customData);
                break;
            }
            case DEVICE_IOS: {
                fireBaseMessage = buildIosMessage(title, body, registrationToken, customData);
                break;
            }
            default: {
                throw new IllegalStateException("Unsupported device type");
            }
        }

        String response;
        try {
            log.debug(String.format("Trying send message to firebase...\ntitle: <%s>\nbody: <%s>\ncustomData: <%s>",
                title, body, customData));
            response = FirebaseMessaging.getInstance().send(fireBaseMessage);
            log.debug("Successfully sent message: " + response);
        } catch (FirebaseMessagingException e) {
            if (ERROR_CODE_TOKEN_NOT_REGISTERED.equals(e.getErrorCode())) {
                log.info("Delete action, because token is not registered. Action:\n" + action);
                actionService.deleteById(action.getId());
            } else {
                log.error("Failed to send firebase message", e);
            }
        }
    }

    private Template getTemplate(String eventType, JsonNode config) {
        String titleTemplate = getDefaultTitleTemplate(eventType);
        String bodyTemplate = getDefaultBodyTemplate(eventType);

        if (config.hasNonNull(TEMPLATE_ID)) {
            String templateId = config.get(TEMPLATE_ID).asText();
            titleTemplate = templateService.getTitleTemplate(templateId);
            bodyTemplate = templateService.getBodyTemplate(templateId);
        } else {
            if (config.hasNonNull(TITLE_TEMPLATE)) {
                titleTemplate = config.get(TITLE_TEMPLATE).asText();
            }

            if (config.hasNonNull(BODY_TEMPLATE)) {
                bodyTemplate = config.get(BODY_TEMPLATE).asText();
            }
        }

        return new Template(titleTemplate, bodyTemplate);
    }

    private Map<String, String> prepareCustomData(EventDto dto) {
        if (!StringUtils.startsWith(dto.resolveType(), "task.")) {
            return new HashMap<>();
        }
        TaskEventDto taskEventDTO = EventDtoFactory.fromEventDto(dto);
        Map<String, String> data = new HashMap<>();

        String taskId = taskEventDTO.getTaskInstanceId();
        if (StringUtils.isNotBlank(taskId)) {
            data.put(PARAM_TASK_ID, taskId);
        }

        String doc = taskEventDTO.getDocument();
        if (StringUtils.isNotBlank(doc)) {
            data.put(PARAM_DOCUMENT, doc);
        }

        return data;
    }

    private String getDefaultTitleTemplate(String eventType) {
        switch (TaskEventType.resolve(eventType)) {
            case ASSIGN:
                return appProps.getFirebase().getTemplate().getDefaultFirebaseTaskAssignTitle();
            case CREATE:
                return appProps.getFirebase().getTemplate().getDefaultFirebaseTaskCreateTitle();
            case COMPLETE:
                return appProps.getFirebase().getTemplate().getDefaultFirebaseTaskCompleteTitle();
            case DELETE:
                return appProps.getFirebase().getTemplate().getDefaultFirebaseTaskDeleteTitle();
            default:
                log.warn(String.format("Event type: <%s> not supported", eventType));
                return "unsupported event type";
        }
    }

    private String getDefaultBodyTemplate(String eventType) {
        switch (TaskEventType.resolve(eventType)) {
            case ASSIGN:
                return appProps.getFirebase().getTemplate().getDefaultFirebaseTaskAssignBody();
            case CREATE:
                return appProps.getFirebase().getTemplate().getDefaultFirebaseTaskCreateBody();
            case COMPLETE:
                return appProps.getFirebase().getTemplate().getDefaultFirebaseTaskCompleteBody();
            case DELETE:
                return appProps.getFirebase().getTemplate().getDefaultFirebaseTaskDeleteBody();
            default:
                log.warn(String.format("Event type: <%s> not supported", eventType));
                return "unsupported event type";
        }
    }

    private Message buildAndroidMessage(String title, String body, String registrationToken, Map<String, String> data) {
        return Message.builder()
            .setNotification(new Notification(
                title,
                body))
            .setAndroidConfig(AndroidConfig.builder()
                .setTtl(3600 * 1000)
                .setNotification(AndroidNotification.builder()
                    .setIcon("stock_ticker_update")
                    .setColor("#f45342")
                    .build())
                .build())
            .setApnsConfig(ApnsConfig.builder()
                .setAps(Aps.builder()
                    .setBadge(42)
                    .build())
                .build())
            .putAllData(data)
            .setToken(registrationToken)
            .build();
    }

    private Message buildIosMessage(String title, String body, String registrationToken, Map<String, String> data) {
        return Message.builder()
            .setNotification(new Notification(
                title,
                body))
            .setApnsConfig(ApnsConfig.builder()
                .setAps(Aps.builder()
                    .setBadge(42)
                    .build())
                .build())
            .putAllData(data)
            .setToken(registrationToken)
            .build();
    }

    @Getter
    private static class Template {

        Template(String title, String body) {
            this.title = title;
            this.body = body;
        }

        private String title;
        private String body;
    }
}

File: src/main/java/ru/citeck/ecos/notifications/service/processors/FirebaseNotificationProcessor.kt
package ru.citeck.ecos.notifications.service.processors

import com.rabbitmq.client.Delivery
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import ru.citeck.ecos.commons.data.ObjectData
import ru.citeck.ecos.context.lib.auth.AuthContext
import ru.citeck.ecos.events.data.dto.EventDto
import ru.citeck.ecos.events.data.dto.pasrse.EventDtoFactory
import ru.citeck.ecos.events.data.dto.task.TaskEventDto
import ru.citeck.ecos.events.data.dto.task.TaskEventType
import ru.citeck.ecos.notifications.config.ApplicationProperties
import ru.citeck.ecos.notifications.domain.firebase.*
import ru.citeck.ecos.notifications.domain.subscribe.repo.ActionEntity
import ru.citeck.ecos.notifications.domain.subscribe.service.ActionService
import ru.citeck.ecos.notifications.domain.template.service.NotificationTemplateService
import ru.citeck.ecos.notifications.lib.Notification
import ru.citeck.ecos.notifications.lib.NotificationType
import ru.citeck.ecos.notifications.lib.service.NotificationService
import ru.citeck.ecos.webapp.api.entity.EntityRef
import java.io.IOException

/**
 *  //TODO: Make the class universal, at the moment it is possible to send notifications only about the TaskEvent event
 *
 * @author Roman Makarskiy
 */

@Component
class FirebaseNotificationProcessor(
    val notificationService: NotificationService,
    val actionService: ActionService,
    val notificationTemplateService: NotificationTemplateService,
    val appProps: ApplicationProperties
) : ActionProcessor() {

    private val log = KotlinLogging.logger {}

    companion object {
        private const val WORKSPACE_SPACES_STORE = "workspace://SpacesStore/"
        private const val ALFRESCO_APP = "alfresco"
        private const val WF_TASK_SOURCE_ID = "wftask"
    }

    init {
        id = "FIREBASE_NOTIFICATION"
    }

    @Value("\${notifications.default.locale}")
    private lateinit var defaultLocale: String

    override fun processImpl(message: Delivery, dto: EventDto, action: ActionEntity, model: Map<String, Any>) {
        AuthContext.runAsSystem {
            sendFirebaseNotification(dto, action)
        }
    }

    private fun sendFirebaseNotification(dto: EventDto, action: ActionEntity) {
        log.debug("FirebaseNotificationProcessor Process start. DTO: \n$dto")

        val config = getValidConfig(action)
        if (config == null) {
            log.error(
                "Firebase notification not send, because config is not valid. " +
                    "Action ${action.id} will be deleted."
            )
            actionService.deleteById(action.id)
            return
        }

        val notificationTransformer = NotificationTransformer(dto, config, action)

        val templateRef = notificationTransformer.template()
        val templateWithMeta = notificationTemplateService.findById(templateRef.getLocalId())
        if (!templateWithMeta.isPresent) {
            log.error(
                "Firebase notification not send, because template with id " +
                    "$templateRef not found. " +
                    "Action ${action.id} will be deleted."
            )
            actionService.deleteById(action.id)
            return
        }

        val notification = Notification.Builder()
            .record(notificationTransformer.record())
            .templateRef(notificationTransformer.template())
            .notificationType(NotificationType.FIREBASE_NOTIFICATION)
            .recipients(notificationTransformer.recipients())
            .additionalMeta(notificationTransformer.additionalMeta())
            .lang(notificationTransformer.locale())
            .build()

        log.debug { "Build notification: \n$notification" }

        notificationService.send(notification)
    }

    private fun getValidConfig(action: ActionEntity): ObjectData? {
        val config = try {
            ObjectData.create(action.configJSON)
        } catch (e: IOException) {
            log.error("Failed read json from string. Action: $action", e)
            return null
        }

        val regToken = config.get(FIREBASE_CONFIG_CLIENT_REG_TOKEN).asText()
        if (regToken.isBlank()) {
            log.error("Token is empty. Action: $action")
            return null
        }

        val deviceType = config.get(FIREBASE_CONFIG_DEVICE_TYPE_KEY).asText()
        if (deviceType.isBlank()) {
            log.error("Device type is empty. Action: $action")
            return null
        }

        return config
    }

    inner class NotificationTransformer(eventDto: EventDto, config: ObjectData, action: ActionEntity) {

        private val taskEventDto: TaskEventDto

        init {
            taskEventDto = EventDtoFactory.fromEventDto(eventDto)
        }

        val record = fun(): EntityRef {
            if (taskEventDto.document.isNullOrBlank()) {
                return EntityRef.EMPTY
            }

            val ref = EntityRef.valueOf(taskEventDto.document)
            if (ref.getAppName().isBlank() && ref.getLocalId().startsWith(WORKSPACE_SPACES_STORE)) {
                return EntityRef.create(ALFRESCO_APP, "", ref.getLocalId())
            }

            return ref
        }

        val task = fun(): EntityRef {
            if (taskEventDto.taskInstanceId.isNullOrBlank()) {
                return EntityRef.EMPTY
            }

            val ref = EntityRef.valueOf(taskEventDto.taskInstanceId)
            val appName = ref.getAppName().ifBlank { ALFRESCO_APP }
            val sourceId = ref.getSourceId().ifBlank { WF_TASK_SOURCE_ID }

            return EntityRef.create(appName, sourceId, ref.getLocalId())
        }

        val template = fun(): EntityRef {
            config.get(FIREBASE_CONFIG_TEMPLATE_ID).asText().let { templateId ->
                return if (templateId.isBlank()) {
                    getDefaultTemplate(eventDto.resolveType())
                } else {
                    EntityRef.valueOf(templateId)
                }
            }
        }

        val locale = fun(): String {
            return config.get(FIREBASE_CONFIG_LOCALE).asText().ifBlank {
                defaultLocale
            }
        }

        val recipients = fun(): Set<String> {
            config.get(FIREBASE_CONFIG_CLIENT_REG_TOKEN).asText().let { regToken ->
                return if (regToken.isBlank()) {
                    emptySet()
                } else {
                    setOf(regToken)
                }
            }
        }

        val additionalMeta = fun(): Map<String, Any> {
            val firebaseMessageData = mutableMapOf<String, String>()

            firebaseMessageData[FIREBASE_MESSAGE_DATA_TASK_ID] = task().toString()
            firebaseMessageData[FIREBASE_MESSAGE_DATA_DOCUMENT] = record().toString()

            val notificationData = mutableMapOf<String, Any>()

            notificationData[FIREBASE_MESSAGE_DATA_KEY] = firebaseMessageData
            notificationData[FIREBASE_CONFIG_DEVICE_TYPE_KEY] = config.get(FIREBASE_CONFIG_DEVICE_TYPE_KEY).asText()
            notificationData[FIREBASE_ACTION_ENTITY_ID_KEY] = action.id

            return mutableMapOf(
                FIREBASE_NOTIFICATION_DATA_KEY to notificationData,
                FIREBASE_NOTIFICATION_TASK_KEY to task()
            )
        }

        private fun getDefaultTemplate(eventType: String): EntityRef {
            return when (TaskEventType.resolve(eventType)) {
                TaskEventType.ASSIGN -> EntityRef.valueOf(appProps.firebase.template.defaultTaskAssignTemplate)
                TaskEventType.CREATE -> EntityRef.valueOf(appProps.firebase.template.defaultTaskCreateTemplate)
                TaskEventType.COMPLETE -> EntityRef.valueOf(appProps.firebase.template.defaultTaskCompleteTemplate)
                TaskEventType.DELETE -> EntityRef.valueOf(appProps.firebase.template.defaultTaskDeleteTemplate)
                else -> {
                    throw EcosFirebaseNotificationException("Event type <$eventType> not supported")
                }
            }
        }
    }
}
